File: lib_local/gallery/src/main/java/com/yanzhenjie/album/mvp/Source.java
package com.yanzhenjie.album.mvp;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.Toolbar;

/**
 * Created by YanZhenjie on 2017/12/8.
 */
abstract class Source<Host> {

    private Host mHost;

    public Source(Host host) {
        mHost = host;
    }

    public Host getHost() {
        return mHost;
    }

    abstract void prepare();

    abstract void setActionBar(Toolbar actionBar);

    abstract MenuInflater getMenuInflater();

    abstract Menu getMenu();

    abstract void setMenuClickListener(MenuClickListener selectedListener);

    abstract void setDisplayHomeAsUpEnabled(boolean showHome);

    abstract void setHomeAsUpIndicator(@DrawableRes int icon);

    abstract void setHomeAsUpIndicator(Drawable icon);

    abstract void setTitle(CharSequence title);

    abstract void setTitle(@StringRes int title);

    abstract void setSubTitle(CharSequence title);

    abstract void setSubTitle(@StringRes int title);

    abstract Context getContext();

    abstract View getView();

    abstract void closeInputMethod();

    interface MenuClickListener {

        void onHomeClick();

        void onMenuClick(MenuItem item);
    }

}

File: lib_local/gallery/src/main/java/com/example/gallery/base/source/Source.kt
package com.example.gallery.base.source

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.widget.Toolbar

/**
 * Created by YanZhenjie on 2017/12/8.
 */
abstract class Source<Host>(private val mHost: Host) {

    fun getHost(): Host? {
        return mHost
    }

    abstract fun prepare()

    abstract fun setActionBar(actionBar: Toolbar?)

    abstract fun getMenuInflater(): MenuInflater?

    abstract fun getMenu(): Menu?

    abstract fun setMenuClickListener(selectedListener: MenuClickListener?)

    abstract fun setDisplayHomeAsUpEnabled(showHome: Boolean)

    abstract fun setHomeAsUpIndicator(@DrawableRes icon: Int)

    abstract fun setHomeAsUpIndicator(icon: Drawable?)

    abstract fun setTitle(title: CharSequence?)

    abstract fun setTitle(@StringRes title: Int)

    abstract fun setSubTitle(title: CharSequence?)

    abstract fun setSubTitle(@StringRes title: Int)

    abstract fun getContext(): Context?

    abstract fun getView(): View?

    abstract fun closeInputMethod()

    interface MenuClickListener {
        /**
         * 左侧返回按钮点击
         */
        fun onHomeClick()

        /**
         * 右侧菜单点击
         */
        fun onMenuClick(item: MenuItem?)
    }

}